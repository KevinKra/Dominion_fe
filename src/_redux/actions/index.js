export const updateTableCards = cards => {
  return {
    type: "UPDATE_TABLE_CARDS",
    cards
  };
};

export const updatePlayerCards = (playerDeck, playerHand, discardPile) => {
  return {
    type: "UPDATE_PLAYER_CARDS",
    playerDeck,
    playerHand,
    discardPile
  };
};

export const beginTurn = () => {
  return {
    type: "START_TURN"
  };
};

export const endTurn = () => {
  return {
    type: "END_TURN"
  };
};

export const cycleToPhase = phase => {
  return {
    type: "CYCLE_PHASE",
    phase
  };
};

export const initiatePlayer = (gameID, playerID, playerName) => {
  return {
    type: "INITIATE_PLAYER",
    gameID,
    playerID,
    playerName
  };
};

export const addTreasure = increase => {
  return {
    type: "ADD_TREASURE",
    increase
  };
};

export const spendTreasure = decrease => {
  return {
    type: "SPEND_TREASURE",
    decrease
  };
};

export const activateCard = card => {
  return {
    type: "ACTIVATE_CARD",
    card
  };
};

export const applyActionValues = (
  spendingPower,
  buyingPower,
  actionsProvided
) => {
  return {
    type: "APPLY_ACTION_VALUES",
    spendingPower,
    buyingPower,
    actionsProvided
  };
};
