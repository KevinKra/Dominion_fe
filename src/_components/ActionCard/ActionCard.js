import React from "react";
// import "./ActionCard.scss";
import * as actions from "../../_redux/actions";
import Card from "../Card/Card";
import { connect } from "react-redux";

export class ActionCard {
  draw = drawCount => {
    const { updatePlayerCards, playerDeck, playerHand, discardPile } = this.props;
    const drawnCards = playerDeck.splice(0, drawCount);
    const newHand = [...playerHand, ...drawnCards];
    updatePlayerCards(playerDeck, newHand, discardPile);
  };

  playActionCard = event => {
    event.preventDefault();
    if (!this.props.playerTurn.isActive) return;
    const {
      spendingPower,
      buyingPower,
      actionsProvided,
      cardsToDraw,
      applyActionValues,
      activateCard,
      playerHand
    } = this.props;
    applyActionValues(spendingPower, buyingPower, actionsProvided);
    cardsToDraw > 0 && this.draw(cardsToDraw);
    activateCard(playerHand.find(card => card.id === event.target.id));
  };

  render() {
    const { card } = this.props;
    const actionCard = (
      <Card
        name={card.name}
        desc={card.desc}
        tags={card.tags}
        category={card.category}
        image={card.image}
        cost={card.cost}
        id={card.id}
        key={card.id}
      />
    );
    return (
      <section
        className='ActionCard'
        id={card.id}
        onClick={event => this.playActionCard(event)}
      >
        {actionCard}
      </section>
    );
  }
}

const mapStateToProps = store => ({
  playerDeck: store.playerDeck,
  playerHand: store.playerHand,
  discardPile: store.discardPile,
  playerTurn: store.playerTurn
});

const mapDispatchToProps = dispatch => ({
  applyActionValues: (spendingPower, buyingPower, actionsProvided) =>
    dispatch(actions.applyActionValues(spendingPower, buyingPower, actionsProvided)),
  updatePlayerCards: (deck, hand, discard) =>
    dispatch(actions.updatePlayerCards(deck, hand, discard)),
  activateCard: card => dispatch(actions.activateCard(card))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ActionCard);
