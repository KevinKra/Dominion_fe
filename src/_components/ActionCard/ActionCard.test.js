import React from "react";
import { shallow } from "enzyme";
import { ActionCard, mapStateToProps, mapDispatchToProps } from "./ActionCard";
import * as actions from "../../_redux/actions";

describe("ActionCard", () => {
  let wrapper, instance, props;
  beforeEach(() => {
    props = {
      card: {
        name: "card",
        desc: "desc",
        tags: [],
        category: "cat",
        image: "..link",
        cost: 0,
        id: 1,
        key: 1
      }
    };
    wrapper = shallow(<ActionCard {...props} />);
    instance = wrapper.instance();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should be able to draw cards", () => {
    instance.draw = jest.fn();
    instance.draw(2);
    expect(instance.draw).toHaveBeenCalledWith(2);
  });

  it("should be able to play action cards on click", () => {
    instance.playActionCard = jest.fn();
    wrapper.find(".ActionCard").simulate("click");
    expect(instance.playActionCard).toHaveBeenCalled();
  });
});

describe("mapStateToProps", () => {
  it("should return the player deck, hand, discardPile, and turn object", () => {
    const mockState = {
      playerDeck: [{ id: 1 }, { id: 2 }],
      playerHand: [],
      discardPile: [],
      playerTurn: { isActive: false, phase: "Standby" },
      uselessPropOne: "example",
      uselessPropTwo: "secondExample"
    };
    const expected = {
      playerDeck: [{ id: 1 }, { id: 2 }],
      playerHand: [],
      discardPile: [],
      playerTurn: { isActive: false, phase: "Standby" }
    };

    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected);
  });
});

describe("mapDispatchToProps", () => {
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = jest.fn();
  });

  it("should call dispatch with spending, buying, and action values when applyActionValues is called", () => {
    const actionToDispatch = actions.applyActionValues(1, 2, 3);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.applyActionValues(1, 2, 3);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it("should call dispatch with deck, hand, and discard when updatePlayerCards is called", () => {
    const actionToDispatch = actions.updatePlayerCards(
      [1, 2, 3],
      [4, 5, 6],
      []
    );
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.updatePlayerCards([1, 2, 3], [4, 5, 6], []);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it("should call dispatch with a card when activateCard is called", () => {
    const actionToDispatch = actions.activateCard({ id: 1 });
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.activateCard({ id: 1 });
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});
