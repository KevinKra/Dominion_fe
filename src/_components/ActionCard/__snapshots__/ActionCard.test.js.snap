// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionCard should match snapshot 1`] = `
<section
  className="ActionCard"
  id={1}
  onClick={[Function]}
>
  <Card
    category="cat"
    cost={0}
    desc="desc"
    id={1}
    image="..link"
    key="1"
    name="card"
    tags={Array []}
  />
</section>
`;
